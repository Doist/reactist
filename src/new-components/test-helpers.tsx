/* eslint-disable jest/no-export */
/* eslint-disable jest/valid-title */
import * as React from 'react'
import { render, screen } from '@testing-library/react'
import { ResponsiveProp } from './responsive-props'
import { Space } from './common-types'

type PropsWithSpace = { space?: ResponsiveProp<Space>; 'data-testid'?: string }

const spaceValues: Array<Space> = ['xsmall', 'small', 'medium', 'large', 'xlarge', 'xxlarge']
const testCases = spaceValues.map((space) => [space])

function getSpaceClassNames(element: HTMLElement) {
    return Array.from(element.classList).filter((className) => className.includes('space-'))
}

function runSpaceTests<Props extends PropsWithSpace>(Component: React.ComponentType<Props>) {
    function renderTestCase(space: ResponsiveProp<Space>) {
        // @ts-expect-error not sure how to properly type the Component argument above
        render(<Component data-testid="subject" space={space} />)
        return screen.getByTestId('subject')
    }

    describe('space', () => {
        test.each(testCases)('it applies the styles needed for space="%s"', (space) => {
            const subject = renderTestCase(space)
            expect(getSpaceClassNames(subject)).toEqual([`space-${space}`])
        })

        it('allows to specify different spacing rules for different screen sizes', () => {
            const subject = renderTestCase(['small', 'medium', 'large'])
            expect(getSpaceClassNames(subject)).toEqual([
                'space-small',
                'tablet-space-medium',
                'desktop-space-large',
            ])
        })
    })
}

export { runSpaceTests }
