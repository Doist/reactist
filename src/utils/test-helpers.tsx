/* eslint-disable jest/no-export */
/* eslint-disable jest/valid-title */
import * as React from 'react'
import { act, render, screen } from '@testing-library/react'
import type { ResponsiveProp } from './responsive-props'
import type { Space } from './common-types'

type PropsWithSpace = { space?: ResponsiveProp<Space>; 'data-testid'?: string }

const spaceValues: Array<Space> = ['xsmall', 'small', 'medium', 'large', 'xlarge', 'xxlarge']
const testCases = spaceValues.map((space) => [space])

function getSpaceClassNames(element: HTMLElement) {
    return Array.from(element.classList).filter((className) => className.includes('gap-'))
}

function runSpaceTests<Props extends PropsWithSpace>(Component: React.ComponentType<Props>) {
    function renderTestCase(space: ResponsiveProp<Space>) {
        // @ts-expect-error not sure how to properly type the Component argument above
        render(<Component data-testid="subject" space={space} />)
        return screen.getByTestId('subject')
    }

    describe('space', () => {
        test.each(testCases)('it applies the styles needed for space="%s"', (space) => {
            const subject = renderTestCase(space)
            expect(getSpaceClassNames(subject)).toEqual([`gap-${space}`])
        })

        it('allows to specify different spacing rules for different screen sizes', () => {
            const subject = renderTestCase({ mobile: 'small', tablet: 'medium', desktop: 'large' })
            expect(getSpaceClassNames(subject)).toEqual([
                'gap-small',
                'tablet-gap-medium',
                'desktop-gap-large',
            ])
        })
    })
}

/**
 * Solves some issues with unwanted warnings in tests of ariakit components due to its internal
 * usage of the event queue for asynchronous side-effects.
 *
 * Think of it as a special version of `act` that we need to call to make sure some async (but
 * immediate) actions are taken care of. Mostly around the ariakit popover and combobox elements'
 * state management.
 *
 * @see https://twitter.com/diegohaz/status/1560525455383461888
 * @see https://github.com/ariakit/ariakit/issues/1800#issuecomment-1227862399
 */
function flushMicrotasks() {
    return act(() => Promise.resolve())
}

function TestIcon() {
    return (
        <svg
            width="24"
            height="24"
            fill="none"
            xmlns="http://www.w3.org/2000/svg"
            aria-hidden
            data-testid="test-icon"
        >
            <path
                fillRule="evenodd"
                clipRule="evenodd"
                d="M12 21a9 9 0 1 0 0-18 9 9 0 0 0 0 18zm0-1a8 8 0 1 1 0-16 8 8 0 0 1 0 16zm1.18-11.84a.84.84 0 1 1-1.68 0 .84.84 0 0 1 1.68 0zM12.5 10a.5.5 0 0 1 .5.5V15h1a.5.5 0 0 1 0 1h-3a.5.5 0 0 1 0-1h1v-4h-1a.5.5 0 0 1 0-1h1.5z"
                fill="currentColor"
            />
        </svg>
    )
}

export { runSpaceTests, flushMicrotasks, TestIcon }
