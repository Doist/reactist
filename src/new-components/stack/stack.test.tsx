import * as React from 'react'
import { render, screen } from '@testing-library/react'
import { runSpaceTests } from '../test-helpers'
import { Stack } from './'

describe('Stack', () => {
    it('does not acknowledge the className prop, but exceptionallySetClassName instead', () => {
        render(
            <Stack
                data-testid="stack"
                // @ts-expect-error
                className="wrong"
                exceptionallySetClassName="right"
            />,
        )
        expect(screen.getByTestId('stack')).toHaveClass('right')
        expect(screen.getByTestId('stack')).not.toHaveClass('wrong')
    })

    it('can be rendered as any HTML element', () => {
        render(<Stack data-testid="stack" as="nav" />)
        expect(screen.getByTestId('stack').tagName).toBe('NAV')
    })

    it('renders its children as its content', () => {
        render(
            <Stack data-testid="stack">
                <div>one</div>
                <div>two</div>
            </Stack>,
        )
        expect(screen.getByTestId('stack').innerHTML).toMatchInlineSnapshot(
            `"<div>one</div><div>two</div>"`,
        )
    })

    it('renders dividers when instructed to do so', () => {
        const { rerender } = render(
            <Stack data-testid="stack" dividers>
                <div>one</div>
                <div>two</div>
                <div>three</div>
            </Stack>,
        )
        expect(screen.getByTestId('stack')).toMatchSnapshot('dividers={true}')

        // regular dividers
        rerender(
            <Stack data-testid="stack" dividers="regular">
                <div>one</div>
                <div>two</div>
                <div>three</div>
            </Stack>,
        )
        expect(screen.getByTestId('stack')).toMatchSnapshot('dividers="regular"')

        // strong dividers
        rerender(
            <Stack data-testid="stack" dividers="strong">
                <div>one</div>
                <div>two</div>
                <div>three</div>
            </Stack>,
        )
        expect(screen.getByTestId('stack')).toMatchSnapshot('dividers="strong"')
    })

    it('applies some extra class names corresponding to other layout-related props', () => {
        render(
            <Stack
                data-testid="stack"
                maxWidth="large"
                minWidth="small"
                padding="medium"
                border="standard"
                borderRadius="standard"
                background="highlight"
            />,
        )
        expect(screen.getByTestId('stack')).toHaveClass(
            'box',
            'minWidth-small',
            'maxWidth-large',
            'paddingTop-medium',
            'paddingRight-medium',
            'paddingBottom-medium',
            'paddingLeft-medium',
            'bg-highlight',
            'borderRadius-standard',
            'border-standard',
        )
    })

    describe('align', () => {
        it('allows to align its children to the left, center or right', () => {
            // no explicit alignment
            const { rerender } = render(<Stack data-testid="stack" />)
            expect(screen.getByTestId('stack')).toHaveClass('display-block')
            expect(screen.getByTestId('stack')).not.toHaveClass('display-flex')
            expect(screen.getByTestId('stack')).not.toHaveClass('flexDirection-column')
            expect(screen.getByTestId('stack')).not.toHaveClass('alignItems-center')

            // aligned to the left (same as when there's no explicit alignment)
            rerender(<Stack data-testid="stack" align="left" />)
            expect(screen.getByTestId('stack')).toHaveClass('display-block')
            expect(screen.getByTestId('stack')).not.toHaveClass('display-flex')
            expect(screen.getByTestId('stack')).not.toHaveClass('flexDirection-column')
            expect(screen.getByTestId('stack')).not.toHaveClass('alignItems-center')

            // aligned to the center
            rerender(<Stack data-testid="stack" align="center" />)
            expect(screen.getByTestId('stack')).not.toHaveClass('display-block')
            expect(screen.getByTestId('stack')).toHaveClass('display-flex')
            expect(screen.getByTestId('stack')).toHaveClass('flexDirection-column')
            expect(screen.getByTestId('stack')).toHaveClass('alignItems-center')

            // aligned to the right
            rerender(<Stack data-testid="stack" align="right" />)
            expect(screen.getByTestId('stack')).not.toHaveClass('display-block')
            expect(screen.getByTestId('stack')).toHaveClass('display-flex')
            expect(screen.getByTestId('stack')).toHaveClass('flexDirection-column')
            expect(screen.getByTestId('stack')).toHaveClass('alignItems-flexEnd')
        })

        it('supports specifying alignment based on viewport size', () => {
            render(<Stack data-testid="stack" align={['left', 'center', 'right']} />)
            expect(screen.getByTestId('stack')).toHaveClass('display-flex')
            expect(screen.getByTestId('stack')).toHaveClass('flexDirection-column')
            expect(screen.getByTestId('stack')).toHaveClass('alignItems-flexStart')
            expect(screen.getByTestId('stack')).toHaveClass('tablet-alignItems-center')
            expect(screen.getByTestId('stack')).toHaveClass('desktop-alignItems-flexEnd')
        })
    })

    runSpaceTests(Stack)
})
