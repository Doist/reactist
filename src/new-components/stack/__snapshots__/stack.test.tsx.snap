// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack renders dividers when instructed to do so: dividers="regular" 1`] = `
<div
  class="box display-block"
  data-testid="stack"
>
  <div>
    one
  </div>
  <hr
    class="divider box display-block"
  />
  <div>
    two
  </div>
  <hr
    class="divider box display-block"
  />
  <div>
    three
  </div>
</div>
`;

exports[`Stack renders dividers when instructed to do so: dividers="strong" 1`] = `
<div
  class="box display-block"
  data-testid="stack"
>
  <div>
    one
  </div>
  <hr
    class="divider weight-strong box display-block"
  />
  <div>
    two
  </div>
  <hr
    class="divider weight-strong box display-block"
  />
  <div>
    three
  </div>
</div>
`;

exports[`Stack renders dividers when instructed to do so: dividers={true} 1`] = `
<div
  class="box display-block"
  data-testid="stack"
>
  <div>
    one
  </div>
  <hr
    class="divider box display-block"
  />
  <div>
    two
  </div>
  <hr
    class="divider box display-block"
  />
  <div>
    three
  </div>
</div>
`;
